src/scion/linearisation/grid.hpp

```diff
@@ -0,0 +1,79 @@
+#ifndef NJOY_SCION_LINEARISATION_GRID
+#define NJOY_SCION_LINEARISATION_GRID
+
+// system includes
+
+// other includes
+
+namespace njoy {
+namespace scion {
+namespace linearisation {
+
+  /**
+   *  @brief Determine the minimum grid for linearisation
+   *
+   *  This template function returns the initial grid for linearisation which
+   *  consists of the following unique set of values: boundaries of the domain,
+   *  the real roots of the first and second derivative in the domain.
+   *
+   *  The FirstDerivative and SecondDerivative objects must have a roots()
+   *  method that returns the real roots of the function.
+   *
+   *  @param coefficients   the coefficients of the Legendre series (from
+   *                        lowest to highest order coefficient)
+   */
+  template < typename X, typename Functor,
+             typename FirstDerivative, typename SecondDerivative >
+  std::vector< X > grid( Functor&& function,
+                         FirstDerivative&& first, SecondDerivative&& second,
+                         const X& lowerLimit, const X& upperLimit ) {
+
+    std::vector< X > grid = { lowerLimit, upperLimit };
+
+    // add roots of the first derivative
+    std::vector< X > temp = first.roots();
+    grid.insert( grid.end(), temp.begin(), temp.end() );
+
+    // add roots of the second derivative
+    temp = second.roots();
+    grid.insert( grid.end(), temp.begin(), temp.end() );
+
+    // sort the grid and remove duplicates
+    std::sort( grid.begin(), grid.end() );
+    grid.erase( std::unique( grid.begin(), grid.end() ), grid.end() );
+
+    // remove everything outside of the domain
+    grid.erase( std::lower_bound( grid.begin(), grid.end(), lowerLimit ), grid.begin() );
```

Comment:
The arguments are inverted here. You aren't erasing the things before the beginning of the domain.